From 3d3bb0bd1a0d60b68fd4dd89e2556461e92029ba Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Mon, 13 Jun 2022 10:33:36 +0300
Subject: [PATCH] [SW-70527] gaudi2: ras: error check in ring funcs

Error return code added in the ring pattern functions.
Incase of expected pattern does not match, 1 as
error value is returned.

This return value will be used to capture BIRA and BIST
failures.

Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
Change-Id: Id6484eb72d3d2187de9ce103e3ab3a183d9352df
---

diff --git a/arch/gaudi2/arc/asic.c b/arch/gaudi2/arc/asic.c
index 5c6fbfe..a2ced03 100644
--- a/arch/gaudi2/arc/asic.c
+++ b/arch/gaudi2/arc/asic.c
@@ -24,6 +24,7 @@
 
 #include "common.h"
 #include <hl.h>
+#include <errno.h>
 
 #define SDRAM_BASE_FULL 0x1000FFFFFD000000ull
 
@@ -197,14 +198,13 @@
 
 	val = regmap_readl(mmPSOC_DFT_APB_WORD1);
 
-	gaudi2_sms_check_pattern(val, ones_mask, zeros_mask);
-
-	return 0;
+	return gaudi2_sms_check_pattern(val, ones_mask, zeros_mask);
 }
 
-static void gaudi2_tpc_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_tpc_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL23, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
@@ -227,7 +227,7 @@
 		udelay(DELAY_7P5K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 	regmap_writel(GAUDI2_RING_PATTERN_VAL33, mmPSOC_DFT_APB_WORD0);
 
@@ -235,19 +235,25 @@
 
 	expected_ones = 0b00000000010100000010100000010100;
 	expected_zeros = 0b00000000001011100001011100001011;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b00001010000001010000001010000001;
 	expected_zeros = 0b10000101110000101110000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b01000000101000000101000000101000;
 	expected_zeros = 0b10111000010111000010111000010111;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST)
 		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
@@ -256,7 +262,9 @@
 
 	expected_ones = 0b00001010000001010000001010000000;
 	expected_zeros = 0b00000101110000101110000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL23, mmPSOC_DFT_APB_WORD0);
@@ -270,19 +278,27 @@
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000100100100100100100100100;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 
 		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000000000010010010010010000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
+
 	}
+
+	return ret;
 }
 
-static void gaudi2_sob_rot_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_sob_rot_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL35, mmPSOC_DFT_APB_WORD0);
 
@@ -300,7 +316,7 @@
 		udelay(DELAY_7P5K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL36, mmPSOC_DFT_APB_WORD0);
@@ -308,7 +324,9 @@
 
 	expected_ones = 0b00000000010100000010100000010100;
 	expected_zeros = 0b00000000001011100001011100001011;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST)
 		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
@@ -317,7 +335,9 @@
 
 	expected_ones = 0b00000000000000000000001010000000;
 	expected_zeros = 0b00000000000000000010000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL35, mmPSOC_DFT_APB_WORD0);
@@ -329,13 +349,18 @@
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000000000000010010010010000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 	}
+
+	return ret;
 }
 
-static void gaudi2_sram_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_sram_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL1, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
@@ -349,7 +374,7 @@
 		udelay(DELAY_5K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL5, mmPSOC_DFT_APB_WORD0);
@@ -360,7 +385,9 @@
 
 	expected_ones = 0b00000000000000000000001010000000;
 	expected_zeros = 0b00000000000000000000000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL1, mmPSOC_DFT_APB_WORD0);
@@ -372,13 +399,18 @@
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000000000000000000000010000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 	}
+
+	return ret;
 }
 
-static void gaudi2_xbar_mid_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_xbar_mid_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL1, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
@@ -392,7 +424,7 @@
 		udelay(DELAY_4K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 	regmap_writel(GAUDI2_RING_PATTERN_VAL5, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIST)
@@ -402,7 +434,9 @@
 
 	expected_ones = 0b00000000000000000000001010000000;
 	expected_zeros = 0b00000000000000000000000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL1, mmPSOC_DFT_APB_WORD0);
@@ -412,13 +446,18 @@
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000000000000000000000010000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 	}
+
+	return ret;
 }
 
-static void gaudi2_vdec_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_vdec_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL8, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
@@ -435,14 +474,16 @@
 		udelay(DELAY_4K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 	regmap_writel(GAUDI2_RING_PATTERN_VAL11, mmPSOC_DFT_APB_WORD0);
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b00000000010100000010100000010100;
 	expected_zeros = 0b00000000001011100001011100001011;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST)
 		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
@@ -451,7 +492,9 @@
 
 	expected_ones = 0b00000000000000000000000000000000;
 	expected_zeros = 0b00000000000000000000000000010000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL8, mmPSOC_DFT_APB_WORD0);
@@ -463,13 +506,18 @@
 
 		expected_ones = 00000000000000000000000000000000;
 		expected_zeros = 00000000000000000000010010010000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 	}
+
+	return ret;
 }
 
-static void gaudi2_sob_xdma_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_sob_xdma_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL38, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
@@ -486,14 +534,16 @@
 		udelay(DELAY_5K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 	regmap_writel(GAUDI2_RING_PATTERN_VAL39, mmPSOC_DFT_APB_WORD0);
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b00000000010100000010100000010100;
 	expected_zeros = 0b00000000001011100001011100001011;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST)
 		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
@@ -502,7 +552,9 @@
 
 	expected_ones = 0b00000000000001010000001010000000;
 	expected_zeros = 0b00000000010000101110000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL38, mmPSOC_DFT_APB_WORD0);
@@ -514,14 +566,18 @@
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000000000010010010010010000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 	}
 
+	return ret;
 }
 
-static void gaudi2_nic_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_nic_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL41, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
@@ -541,7 +597,7 @@
 		udelay(DELAY_7P5K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 	regmap_writel(GAUDI2_RING_PATTERN_VAL42, mmPSOC_DFT_APB_WORD0);
 
@@ -549,13 +605,17 @@
 
 	expected_ones = 0b00000000010100000010100000010100;
 	expected_zeros = 0b00000000001011100001011100001011;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b00001010000001010000001010000001;
 	expected_zeros = 0b10000101110000101110000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST)
 		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
@@ -564,8 +624,9 @@
 
 	expected_ones = 0b00000000000000010000001010000000;
 	expected_zeros = 0b00000000000000101110000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL41, mmPSOC_DFT_APB_WORD0);
@@ -578,19 +639,26 @@
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000100100100100100100100100;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 
 		regmap_writel(GAUDI2_RING_PATTERN_VAL21, mmPSOC_DFT_APB_WORD1);
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000000000000000000000000000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 	}
+
+	return ret;
 }
 
-static void gaudi2_mme_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_mme_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL42, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
@@ -610,20 +678,24 @@
 		udelay(DELAY_7P5K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 	regmap_writel(GAUDI2_RING_PATTERN_VAL43, mmPSOC_DFT_APB_WORD0);
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b00000000010100000010100000010100;
 	expected_zeros = 0b00000000001011100001011100001011;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b00001010000001010000001010000001;
 	expected_zeros = 0b10000101110000101110000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST)
 		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
@@ -632,7 +704,10 @@
 
 	expected_ones = 0b00000010000001010000001010000000;
 	expected_zeros = 0b00000101110000101110000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
+
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL42, mmPSOC_DFT_APB_WORD0);
 		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
@@ -644,19 +719,26 @@
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000100100100100100100100100;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 
 		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000000000000000000000010000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 	}
+
+	return ret;
 }
 
-static void gaudi2_hmmu_xbar_edge_mc_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_hmmu_xbar_edge_mc_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL44, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
@@ -676,26 +758,32 @@
 		udelay(DELAY_7P5K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 	regmap_writel(GAUDI2_RING_PATTERN_VAL48, mmPSOC_DFT_APB_WORD0);
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b00000000010100000010100000010100;
 	expected_zeros = 0b00000000001011100001011100001011;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b00001010000001010000001010000001;
 	expected_zeros = 0b10000101110000101110000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b01000000101000000101000000101000;
 	expected_zeros = 0b10111000010111000010111000010111;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST)
 		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
@@ -704,8 +792,9 @@
 
 	expected_ones = 0b00000000000000000000000000000000;
 	expected_zeros = 0b00000000000000000000000000000000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL44, mmPSOC_DFT_APB_WORD0);
@@ -718,19 +807,26 @@
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000100100100100100100100100;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 
 		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000000000000000000010010000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 	}
+
+	return ret;
 }
 
-static void gaudi2_hmmu_2_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_hmmu_2_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL50, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
@@ -744,7 +840,7 @@
 		udelay(DELAY_7P5K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 	regmap_writel(GAUDI2_RING_PATTERN_VAL51, mmPSOC_DFT_APB_WORD0);
 
@@ -755,7 +851,9 @@
 
 	expected_ones = 0b00000000000001010000001010000000;
 	expected_zeros = 0b00000000000000101110000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL50, mmPSOC_DFT_APB_WORD0);
@@ -766,13 +864,18 @@
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000000000000000000010010000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 	}
+
+	return ret;
 }
 
-static void gaudi2_edma1_hmmu3_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_edma1_hmmu3_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL38, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
@@ -789,14 +892,16 @@
 		udelay(DELAY_7P5K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 	regmap_writel(GAUDI2_RING_PATTERN_VAL39, mmPSOC_DFT_APB_WORD0);
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b00000000010100000010100000010100;
 	expected_zeros = 0b00000000001011100001011100001011;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST)
 		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
@@ -805,7 +910,9 @@
 
 	expected_ones = 0b00000000000001010000001010000000;
 	expected_zeros = 0b00000000010000101110000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL38, mmPSOC_DFT_APB_WORD0);
@@ -817,13 +924,18 @@
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000000000010010010010010000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 	}
+
+	return ret;
 }
 
-static void gaudi2_hmmu_mc_ring_pattern(enum gaudi2_sram_self_repair_op op)
+static int gaudi2_hmmu_mc_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
+	int err, ret = 0;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL53, mmPSOC_DFT_APB_WORD0);
 	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
@@ -843,20 +955,24 @@
 		udelay(DELAY_7P5K_USEC);
 	} else {
 		preboot_puts("Invalid sram repair operation\n");
-		return;
+		return -EINVAL;
 	}
 	regmap_writel(GAUDI2_RING_PATTERN_VAL55, mmPSOC_DFT_APB_WORD0);
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b00000000010100000010100000010100;
 	expected_zeros = 0b00000000001011100001011100001011;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
 
 	expected_ones = 0b00001010000001010000001010000001;
 	expected_zeros = 0b10000101110000101110000101110000;
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST)
 		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
@@ -870,7 +986,9 @@
 		expected_ones = 0b00000000000000000000000010000000;
 		expected_zeros = 0b00000000000000000000000101110000;
 	}
-	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	if (err)
+		ret = 1;
 
 	if (op == SRAM_SELF_REPAIR_OP_BIST) {
 		regmap_writel(GAUDI2_RING_PATTERN_VAL53, mmPSOC_DFT_APB_WORD0);
@@ -883,8 +1001,12 @@
 
 		expected_ones = 0b00000000000000000000000000000000;
 		expected_zeros = 0b00000000010010010010010010010000;
-		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+		if (err)
+			ret = 1;
 	}
+
+	return ret;
 }
 
 static void gaudi2_sram_ring_sel(uint32_t dfx_val, uint32_t ring)
@@ -924,7 +1046,7 @@
 	regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
 }
 
-static void (*gaudi2_sram_region_ring_pattern[])(enum gaudi2_sram_self_repair_op op)  = {
+static int (*gaudi2_sram_region_ring_pattern[])(enum gaudi2_sram_self_repair_op op)  = {
 				[SRAM_RING_TYPE_TPC] = gaudi2_tpc_ring_pattern,
 				[SRAM_RING_TYPE_SOB_ROT] = gaudi2_sob_rot_ring_pattern,
 				[SRAM_RING_TYPE_SRAM] = gaudi2_sram_ring_pattern,
